Bug 	1337539 - Add option to mount EFS on $HOME

fbertsch opened the pull request:
This change will mount EFS on $HOME. The first time a user logs in, the user's EFS directory will be created, along with appropriate directory structure within.

Along with this, I had to move ~/.bashrc code to a "global" .sh file, which runs whenever a user logs in. This way we can keep changing what we put in there without it messing with the user's .bashrc file.

Examples are downloaded on first login, and if EFS is not mounted. With the current implementation, examples will not be updated. We can always add examples later, as a one-off on existing directories, and then adding it to newly created dirs in this code. (Though in general I'm against making changes to a user's dir.)

I've tested this as before:
1. A new user
2. An existing user
3. Without EFS mounting (--email unset)

All three worked as expected. I tested Jupyter and Spark within each, running the longitudinal notebook.

I've requested all of you to review because this is *very* difficult to test, and small changes can affect things over time. I'm sure I missed some things.

fbertsch commented:
Realized I was mounting on worker nodes as well. Could eventually be a use case for this, but right now it's disallowed.

jasonthomas commented on ansible/files/telemetry.sh:
```diff
@@ -1,19 +1,111 @@
+if [ "$SETUP_HOME_DIR" = true ] ; then
+    # Only execute this when using EFS and this is a new user, or when not using EFS
+
+    # Download examples to analyses dir
```
Do these examples change often? Should we replace/update them every time?

fbertsch replied:
This is a discussion we had in the previous PR. I don't want to change *anything* in a user's dir, so they are just going to be there to begin with, and never updated or overwritten. They do not change often. We can kick the can down the road on whether we will want to add more examples later, since we can always just go through and stick them in people's dirs if we feel we really need to.

jasonthomas commented on ansible/files/telemetry.sh:
```diff
@@ -122,34 +186,35 @@ wget -nc https://mran.microsoft.com/install/RRO-3.2.1-el6.x86_64.tar.gz
+
+# Setup global bash file
+# .sh files in /etc/profile.d/ run whenever a user logs in
+GLOBAL_BASHRC=/etc/profile.d/atmo.sh
```
Can we move this out into it's own separate file and wget it?

fbertsch replied:
Hmm. I suppose we could, but we use a whole lot of options in this file ($ANACONDAPATH, memory configs, etc.) which means changes here would need to be paralleled in that new file. I think it makes sense to keep it contained here for now.

jasonthomas commented:
r+wc

fbertsch commented on ansible/files/telemetry.sh:
```diff
@@ -1,19 +1,111 @@
+#EFS Mounting
+if [ -n "$EMAIL" ] && [ -n "$EFS_DNS" ] && [ "$IS_MASTER" = true ]; then
+    # Mount entire efs locally to check for user's dir
+    mkdir -p /mnt/efs
```
Everything seemed to work without it.

mreid-moz commented on ansible/files/telemetry.sh:
```diff
@@ -1,19 +1,111 @@
+    sudo chown --recursive hadoop:hadoop "$HOME"
+
+    # allow Jupyter to access analyses dir
+    sudo chmod --recursive 777 "$HOME/analyses"
```
nit: Do we really need 777 here? Would `chmod a+rw` suffice?

fbertsch replied:
Probably not. I can test out the minimum viable permissions.

mreid-moz commented:
As discussed on IRC, I'd also like to see the EFS volume created in the `ansible/deploy.yml` file.

fbertsch commented:
This is just getting bigger and bigger. I've added the EFS creation in the ansible deploy.yml file. The RRO download is not in /mnt rather than $HOME, since $HOME will be mounted. I've also changed some of the .ssh files, for example, creating the authorized_keys file if it doesn't exist.

fbertsch commented:
Hey all, this has been RFAL.

jezdez commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -62,6 +26,14 @@ while [ $# -gt 0 ]; do
+            ;;
+        --efs-dns)
+            shift
+            EFS_DNS=$1
```
@fbertsch Is this something that we need to add in https://github.com/mozilla/telemetry-analysis-service/issues/217 as well?

fbertsch replied:
@jezdez yes, we'll have to send in the efs-dns. Prod is currently 'fs-d0c30f79.efs.us-west-2.amazonaws.com', dev is 'fs-616ca0c8.efs.us-west-2.amazonaws.com', though I'm not sure how we'd want to differentiate using those.

jezdez replied:
@fbertsch Can you clarify what "dev" means in this instance? ATMO has stage and prod environments, no official dev environment (yet).

Since the most pressing issues with ATMO (and Airflow etc) are blocked by having a [common set of EMR scripts](https://github.com/mozilla/telemetry-analysis-service/issues/203) I'm going to be moving the content of this repo into the one of ATMO. That means I'll be able to hardcode the EFS URIs into ATMO but keep them configurable. I just need to know how the two values you mention map to the dev/stage/prod environments of ATMO.

fbertsch replied:
@jezdez sorry for the delay in answering - it's not totally clear yet. It seems to me that even if someone is launching on the dev version of ATMO, they would still want to use the prod version of EFS. @vitillo, can you chime in here with thoughts you had on separate environments for EFS?

vitillo replied:
I think there should be a different EFS drive per environment to avoid conflicts.

jezdez commented:
@fbertsch I've opened https://github.com/mozilla/emr-bootstrap-spark/pull/60 in case you want to take a look at it

mreid-moz commented:
My concerns are all addressed now. LGTM.

fbertsch commented:
Are we ready to proceed with this change? I need someone to approve the PR.

vitillo commented:
@fbertsch: jason is the right person to review, and approve, this PR.

jasonthomas commented on ansible/deploy.yml:
```diff
@@ -33,3 +33,17 @@
+            app: telemetry-analysis
+        targets:
+            - subnet_id: subnet-63f72706
+              security_groups: [ "sg-0964dc71" ]
```
How are these security groups created today? I generally shy away from hardcoding values. If this security group is created via Cloudformation or something else maybe we can call AWS API to figure out which one to use.

jasonthomas replied:
I think we should create this security group via ansible using the ec2_group module [1]. The output should return the security group id and can be referenced in the efs creation. 

[1] http://docs.ansible.com/ansible/ec2_group_module.html

jasonthomas commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -73,11 +45,103 @@ while [ $# -gt 0 ]; do
+
+CURRENT_KEY=`cat $HOME/.ssh/authorized_keys`
+TELEMETRY_CONF_BUCKET=s3://{{telemetry_analysis_spark_emr_bucket}}
+MEMORY_OVERHEAD=7000  # Tuned for c3.4xlarge
```
We should  consider setting these values based on percentage of TotalMemory so that we can make use of other instance sizes. Although this out of scope for this specific PR and I am not going to block on it.

vitillo replied:
+1

fbertsch replied:
We have an issue for that out [0]. Would be interesting to do some experimentation on that point. Might be worth putting a bug out.

[0] https://github.com/mozilla/telemetry-analysis-service/issues/229

fbertsch commented:
@jasonthomas I've tested this with a new user, with an existing user, and without email. If you see any particular configurations that I should test, (e.g. removing authorized_hosts file for an existing user, which I did test :)) please call it out and I can try it out.

fbertsch commented:
@jasonthomas changed it so the security groups are created as well. RFAL

jasonthomas commented on ansible/deploy.yml:
```diff
@@ -33,3 +33,34 @@
+    - name: EFS security group in region us-west-2
+      local_action:
+        module: ec2_group
+        name: efs-access
```
Let's use a more descriptive name here, like atmo-efs-access or emr-efs-access.

fbertsch commented:
@jasonthomas done.

fbertsch commented:
I say we deploy this Monday. We'll also need to send an email out about the change, but that won't be in effect until https://github.com/mozilla/telemetry-analysis-service/pull/248 is deployed.
